backend/crud.py

```diff
@@ -0,0 +1,211 @@
+from typing import Dict
+from sqlalchemy.orm import Session
+from sqlalchemy import select, update
+
+from .models import (
+    AttrType,
+    Attribute,
+    AttributeDefinition,
+    BoundFK,
+    Entity,
+    Schema,
+    Value,
+    val_map,
+    type_map
+)
+
+from .schemas import (
+    SchemaCreateSchema,
+    SchemaUpdateSchema,
+    AttributeCreateSchema
+)
+
+
+def create_attribute(db: Session, data: AttributeCreateSchema) -> Attribute:
+    exists = db.execute(select(Attribute).where(Attribute.name == data.name)).scalar()
+    if exists:
+        # TODO custom exceptions
+        raise Exception(f'Attribute with name `{data.name}` already exists')
+    a = Attribute(name=data.name, type=data.type)
+    db.add(a)
+    db.commit()
+    return a
+
+
+def create_schema(db: Session, data: SchemaCreateSchema) -> Schema:
+    # TODO require lock here and in update?
+    exists = db.execute(select(Schema).where( (Schema.name == data.name) | (Schema.slug == data.slug) )).scalar()
+    if exists:
+        raise Exception(f'Schema with name `{data.name}` or slug `{data.slug}` already exists')
+    
+    sch = Schema(name=data.name, slug=data.slug)
+    db.add(sch)
+    db.flush()
+    for attr in data.attributes:
+        a: Attribute = db.execute(select(Attribute).where(Attribute.id == attr.attr_id)).scalar()
```

Comment:
If we used the `create_attribute()` function (with the `get_or_create_ equivalent mentioned in an earlier comment), we can indeed drop the following check and exception.

Reply:
> If we used the create_attribute() function (with the `get_or_create_ equivalent mentioned in an earlier comment)

How this would stop me from providing data like
```
{
  ...
 "attributes": [
    {
       attr_id: 99999
       ...
    }
  ]
}
```
when there is no `Attribute` with provided id?

Reply:
Yes, that would require some changes. So the data would need to include the `name` and `type`, so that the attribute object can be created of obtained.

Could that work for you?

Reply:
Well, it should be implementable. I can't say whether it is a good idea to go this way, but we can try at least